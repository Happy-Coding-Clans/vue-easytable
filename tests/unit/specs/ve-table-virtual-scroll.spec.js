import { mount } from "@vue/test-utils";
import veTable from "@/ve-table";
import { later, mockScrollTo } from "../util";
import bodyTrScrolling from "@/ve-table/src/body/body-tr-scrolling.jsx";

describe("veTable virtual scroll", () => {
    // same row height
    let TABLE_DATA_SAME_ROW_HEIGHT = [];
    for (var i = 0; i < 1000; i++) {
        TABLE_DATA_SAME_ROW_HEIGHT.push({
            rowKey: i,
            name: i,
            date: i,
            hobby: i,
            address: i,
        });
    }

    // different row height
    function getRandom(min, max) {
        return Math.floor(Math.random() * (max - min) + min);
    }

    let TABLE_DATA_DIFFERENT_ROW_HEIGHT = [];

    for (var i = 0; i < 1000; i++) {
        let value = i;
        if (i % 2 === 0) {
            const rowCount = getRandom(3, 5);

            for (let i = 0; i < rowCount; i++) {
                value += `this is the long word.<br />`;
            }
        }

        TABLE_DATA_DIFFERENT_ROW_HEIGHT.push({
            rowKey: i,
            name: value,
            date: i,
            hobby: i,
            address: i,
        });
    }

    //
    const MAX_HEIGHT = 500;
    const MIN_ROW_HEIGHT = 40;

    // 表格渲染的数量
    //const TABLE_ROW_COUNT = Math.ceil(MAX_HEIGHT / MIN_ROW_HEIGHT) + 1;
    const TABLE_ROW_COUNT = Math.ceil(MAX_HEIGHT / MIN_ROW_HEIGHT) * 2;

    // get table rendered row count by row height
    function getTableRenderedRowCountByRowHeight(rowHeight, bufferScale = 2) {
        return Math.ceil(MAX_HEIGHT / rowHeight) * bufferScale;
    }

    it("render same row height", async () => {
        const wrapper = mount(veTable, {
            propsData: {
                columns: [
                    {
                        field: "name",
                        key: "b",
                        title: "Name",
                        width: 200,
                        align: "left",
                    },
                    {
                        field: "hobby",
                        key: "c",
                        title: "Hobby",
                        width: 300,
                        align: "left",
                    },
                    {
                        field: "address",
                        key: "d",
                        title: "Address",
                        width: "",
                        align: "left",
                    },
                ],
                tableData: TABLE_DATA_SAME_ROW_HEIGHT,
                virtualScrollOption: {
                    // 是否开启
                    enable: true,
                },
                maxHeight: MAX_HEIGHT,
                rowKeyFieldName: "rowKey",
            },
        });

        wrapper.triggerResizeObserver({ width: MAX_HEIGHT });

        await later();

        expect(wrapper.html()).toMatchSnapshot();
    });

    it("dynamically turns virtual scrolling on", async () => {
        const wrapper = mount(veTable, {
            propsData: {
                columns: [
                    {
                        field: "name",
                        key: "b",
                        title: "Name",
                        width: 200,
                        align: "left",
                    },
                    {
                        field: "hobby",
                        key: "c",
                        title: "Hobby",
                        width: 300,
                        align: "left",
                    },
                    {
                        field: "address",
                        key: "d",
                        title: "Address",
                        width: "",
                        align: "left",
                    },
                ],
                tableData: TABLE_DATA_SAME_ROW_HEIGHT,
                virtualScrollOption: {
                    // 是否开启
                    enable: false,
                },
                maxHeight: MAX_HEIGHT,
                rowKeyFieldName: "rowKey",
            },
        });

        wrapper.triggerResizeObserver({ width: MAX_HEIGHT });

        await later();

        expect(wrapper.find(".ve-table-virtual-scroll").exists()).toBe(false);

        wrapper.setProps({
            virtualScrollOption: {
                // 是否开启
                enable: true,
            },
        });

        await later();

        const tableContainerEl = wrapper.find(".ve-table-container");

        expect(tableContainerEl.attributes("style")).toBe(
            "max-height: 500px; height: 500px;",
        );

        expect(wrapper.find(".ve-table-virtual-scroll").exists()).toBe(true);
        expect(wrapper.find(".ve-table-virtual-phantom").exists()).toBe(true);
        expect(wrapper.findAll(".ve-table-body-tr").length).toBe(
            TABLE_ROW_COUNT,
        );
    });

    it("same row height", async () => {
        const wrapper = mount(veTable, {
            propsData: {
                columns: [
                    {
                        field: "name",
                        key: "b",
                        title: "Name",
                        width: 200,
                        align: "left",
                    },
                    {
                        field: "hobby",
                        key: "c",
                        title: "Hobby",
                        width: 300,
                        align: "left",
                    },
                    {
                        field: "address",
                        key: "d",
                        title: "Address",
                        width: "",
                        align: "left",
                    },
                ],
                tableData: TABLE_DATA_SAME_ROW_HEIGHT,
                virtualScrollOption: {
                    // 是否开启
                    enable: true,
                },
                maxHeight: MAX_HEIGHT,
                rowKeyFieldName: "rowKey",
            },
        });

        wrapper.triggerResizeObserver({ width: MAX_HEIGHT });

        await later();

        expect(wrapper.findAll(".ve-table-body-tr").length).toBe(
            TABLE_ROW_COUNT,
        );
    });

    it("different row height", async () => {
        const wrapper = mount(veTable, {
            propsData: {
                columns: [
                    {
                        field: "name",
                        key: "b",
                        title: "Name",
                        width: 200,
                        align: "left",
                    },
                    {
                        field: "hobby",
                        key: "c",
                        title: "Hobby",
                        width: 300,
                        align: "left",
                    },
                    {
                        field: "address",
                        key: "d",
                        title: "Address",
                        width: "",
                        align: "left",
                    },
                ],
                tableData: TABLE_DATA_DIFFERENT_ROW_HEIGHT,
                virtualScrollOption: {
                    // 是否开启
                    enable: true,
                },
                maxHeight: MAX_HEIGHT,
                rowKeyFieldName: "rowKey",
            },
        });

        wrapper.triggerResizeObserver({ width: MAX_HEIGHT });

        await later();

        expect(wrapper.findAll(".ve-table-body-tr").length).toBe(
            TABLE_ROW_COUNT,
        );
    });

    it("buffer scale", async () => {
        const bufferCount = 10;

        const wrapper = mount(veTable, {
            propsData: {
                columns: [
                    {
                        field: "name",
                        key: "b",
                        title: "Name",
                        width: 200,
                        align: "left",
                    },
                    {
                        field: "hobby",
                        key: "c",
                        title: "Hobby",
                        width: 300,
                        align: "left",
                    },
                    {
                        field: "address",
                        key: "d",
                        title: "Address",
                        width: "",
                        align: "left",
                    },
                ],
                tableData: TABLE_DATA_DIFFERENT_ROW_HEIGHT,
                virtualScrollOption: {
                    // 是否开启
                    enable: true,
                    bufferScale: bufferCount,
                },
                maxHeight: MAX_HEIGHT,
                rowKeyFieldName: "rowKey",
            },
        });

        wrapper.triggerResizeObserver({ width: MAX_HEIGHT });

        await later();

        expect(wrapper.findAll(".ve-table-body-tr").length).toBe(
            Math.ceil(MAX_HEIGHT / MIN_ROW_HEIGHT) * (bufferCount + 1),
        );
    });

    it("with row checkbox", async () => {
        const wrapper = mount(veTable, {
            propsData: {
                columns: [
                    {
                        field: "name",
                        key: "b",
                        title: "Name",
                        width: 200,
                        align: "left",
                        // type=checkbox
                        type: "checkbox",
                    },
                    {
                        field: "hobby",
                        key: "c",
                        title: "Hobby",
                        width: 300,
                        align: "left",
                    },
                    {
                        field: "address",
                        key: "d",
                        title: "Address",
                        width: "",
                        align: "left",
                    },
                ],
                tableData: TABLE_DATA_SAME_ROW_HEIGHT,
                virtualScrollOption: {
                    // 是否开启
                    enable: true,
                },
                checkboxOption: {},
                maxHeight: MAX_HEIGHT,
                rowKeyFieldName: "rowKey",
            },
        });

        wrapper.triggerResizeObserver({ width: MAX_HEIGHT });

        await later();

        expect(wrapper.find(".ve-checkbox").exists()).toBe(true);

        expect(wrapper.findAll(".ve-table-body-tr").length).toBe(
            TABLE_ROW_COUNT,
        );
    });

    it("with row radio", async () => {
        const wrapper = mount(veTable, {
            propsData: {
                columns: [
                    {
                        field: "name",
                        key: "b",
                        title: "Name",
                        width: 200,
                        align: "left",
                        // type=radio
                        type: "radio",
                    },
                    {
                        field: "hobby",
                        key: "c",
                        title: "Hobby",
                        width: 300,
                        align: "left",
                    },
                    {
                        field: "address",
                        key: "d",
                        title: "Address",
                        width: "",
                        align: "left",
                    },
                ],
                tableData: TABLE_DATA_SAME_ROW_HEIGHT,
                virtualScrollOption: {
                    // 是否开启
                    enable: true,
                },
                radioOption: {},
                maxHeight: MAX_HEIGHT,
                rowKeyFieldName: "rowKey",
            },
        });

        wrapper.triggerResizeObserver({ width: MAX_HEIGHT });

        await later();

        expect(wrapper.find(".ve-radio").exists()).toBe(true);

        expect(wrapper.findAll(".ve-table-body-tr").length).toBe(
            TABLE_ROW_COUNT,
        );
    });

    it("with row expand", async () => {
        const wrapper = mount(veTable, {
            propsData: {
                columns: [
                    {
                        field: "name",
                        key: "b",
                        title: "Name",
                        width: 200,
                        align: "left",
                        // type=expand
                        type: "expand",
                    },
                    {
                        field: "hobby",
                        key: "c",
                        title: "Hobby",
                        width: 300,
                        align: "left",
                    },
                    {
                        field: "address",
                        key: "d",
                        title: "Address",
                        width: "",
                        align: "left",
                    },
                ],
                tableData: TABLE_DATA_SAME_ROW_HEIGHT,
                virtualScrollOption: {
                    // 是否开启
                    enable: true,
                },
                expandOption: {
                    render: ({ row, column, rowIndex }, h) => {
                        return (
                            <p>
                                My name is{" "}
                                <span style="color:#1890ff;">{row.name}</span>
                                ,I'm living in {row.address}
                            </p>
                        );
                    },
                },
                maxHeight: MAX_HEIGHT,
                rowKeyFieldName: "rowKey",
            },
        });

        wrapper.triggerResizeObserver({ width: MAX_HEIGHT });

        await later();

        expect(wrapper.find(".ve-table-row-expand-icon").exists()).toBe(true);

        expect(wrapper.findAll(".ve-table-body-tr").length).toBe(
            TABLE_ROW_COUNT,
        );
    });

    it("with column fixed", async () => {
        const wrapper = mount(veTable, {
            propsData: {
                columns: [
                    {
                        field: "name",
                        key: "b",
                        title: "Name",
                        width: 200,
                        align: "left",
                        fixed: "left",
                    },
                    {
                        field: "hobby",
                        key: "c",
                        title: "Hobby",
                        width: 300,
                        align: "left",
                    },
                    {
                        field: "address",
                        key: "d",
                        title: "Address",
                        width: "",
                        align: "left",
                        fixed: "right",
                    },
                ],
                tableData: TABLE_DATA_SAME_ROW_HEIGHT,
                virtualScrollOption: {
                    // 是否开启
                    enable: true,
                },
                maxHeight: MAX_HEIGHT,
                rowKeyFieldName: "rowKey",
            },
        });

        wrapper.triggerResizeObserver({ width: MAX_HEIGHT });

        await later();

        const thEls = wrapper.findAll(
            ".ve-table-header .ve-table-header-tr .ve-table-header-th",
        );

        expect(thEls.at(0).classes()).toContain("ve-table-fixed-left");
        expect(thEls.at(0).classes()).toContain(
            "ve-table-last-left-fixed-column",
        );

        expect(thEls.at(2).classes()).toContain("ve-table-fixed-right");
        expect(thEls.at(2).classes()).toContain(
            "ve-table-first-right-fixed-column",
        );

        const tdEls = wrapper.findAll(
            ".ve-table-header .ve-table-header-tr .ve-table-header-th",
        );

        expect(tdEls.at(0).classes()).toContain("ve-table-fixed-left");
        expect(tdEls.at(0).classes()).toContain(
            "ve-table-last-left-fixed-column",
        );

        expect(tdEls.at(2).classes()).toContain("ve-table-fixed-right");
        expect(tdEls.at(2).classes()).toContain(
            "ve-table-first-right-fixed-column",
        );

        expect(wrapper.findAll(".ve-table-body-tr").length).toBe(
            TABLE_ROW_COUNT,
        );
    });

    it("with column footer", async () => {
        const wrapper = mount(veTable, {
            propsData: {
                columns: [
                    {
                        field: "name",
                        key: "b",
                        title: "Name",
                        width: 200,
                        align: "left",
                        fixed: "left",
                    },
                    {
                        field: "hobby",
                        key: "c",
                        title: "Hobby",
                        width: 300,
                        align: "left",
                    },
                    {
                        field: "address",
                        key: "d",
                        title: "Address",
                        width: "",
                        align: "left",
                        fixed: "right",
                    },
                ],
                tableData: TABLE_DATA_SAME_ROW_HEIGHT,
                footerData: [
                    {
                        rowKey: 0,
                        name: "平均值",
                        date: 1100,
                        hobby: 1200,
                        address: 1300,
                    },
                    {
                        rowKey: 1,
                        name: "汇总值",
                        date: 701000,
                        hobby: 801000,
                        address: 801000,
                    },
                ],
                virtualScrollOption: {
                    // 是否开启
                    enable: true,
                },
                maxHeight: MAX_HEIGHT,
                rowKeyFieldName: "rowKey",
            },
        });

        wrapper.triggerResizeObserver({ width: MAX_HEIGHT });

        await later();

        expect(
            wrapper.findAll(".ve-table-footer-tr .ve-table-footer-td").length,
        ).toBe(6);

        expect(wrapper.findAll(".ve-table-body-tr").length).toBe(
            TABLE_ROW_COUNT,
        );
    });

    it("scrolling callback", async () => {
        const mockFn = jest.fn();
        const wrapper = mount(veTable, {
            propsData: {
                columns: [
                    {
                        field: "name",
                        key: "b",
                        title: "Name",
                        width: 200,
                        align: "left",
                    },
                    {
                        field: "hobby",
                        key: "c",
                        title: "Hobby",
                        width: 300,
                        align: "left",
                    },
                    {
                        field: "address",
                        key: "d",
                        title: "Address",
                        width: "",
                        align: "left",
                    },
                ],
                tableData: TABLE_DATA_SAME_ROW_HEIGHT,
                virtualScrollOption: {
                    // 是否开启
                    enable: true,
                    scrolling: mockFn,
                },
                maxHeight: MAX_HEIGHT,
                rowKeyFieldName: "rowKey",
            },
        });

        wrapper.triggerResizeObserver({ width: MAX_HEIGHT });

        await later();

        expect(mockFn).toHaveBeenCalledTimes(1);

        expect(mockFn).toHaveBeenCalledWith({
            startRowIndex: 0,
            visibleStartIndex: 0,
            visibleEndIndex: Math.ceil(MAX_HEIGHT / MIN_ROW_HEIGHT),
            visibleAboveCount: 0,
            visibleBelowCount: Math.ceil(MAX_HEIGHT / MIN_ROW_HEIGHT),
        });
    });

    it("minRowHeight prop", async () => {
        const mockFn = jest.fn();
        const minRowHeight = 50;
        const wrapper = mount(veTable, {
            propsData: {
                columns: [
                    {
                        field: "name",
                        key: "b",
                        title: "Name",
                        width: 200,
                        align: "left",
                    },
                    {
                        field: "hobby",
                        key: "c",
                        title: "Hobby",
                        width: 300,
                        align: "left",
                    },
                    {
                        field: "address",
                        key: "d",
                        title: "Address",
                        width: "",
                        align: "left",
                    },
                ],
                tableData: TABLE_DATA_SAME_ROW_HEIGHT,
                virtualScrollOption: {
                    // 是否开启
                    enable: true,
                    minRowHeight: minRowHeight,
                },
                maxHeight: MAX_HEIGHT,
                rowKeyFieldName: "rowKey",
            },
        });

        wrapper.triggerResizeObserver({ width: MAX_HEIGHT });

        await later();

        const rowCount = getTableRenderedRowCountByRowHeight(minRowHeight);
        expect(wrapper.findAll(".ve-table-body-tr").length).toBe(rowCount);
    });

    /* 
    how to test?
    */
    it("virtual scroll palceholder", async () => {
        const wrapper = mount(bodyTrScrolling, {
            propsData: {
                colgroups: [{}],
            },
        });

        expect(wrapper.html()).toMatchSnapshot();
    });

    // it("virtual scroll placeholder", async () => {
    //     const wrapper = mount(veTable, {
    //         propsData: {
    //             columns: [
    //                 {
    //                     field: "name",
    //                     key: "b",
    //                     title: "Name",
    //                     width: 200,
    //                     align: "left",
    //                 },
    //                 {
    //                     field: "hobby",
    //                     key: "c",
    //                     title: "Hobby",
    //                     width: 300,
    //                     align: "left",
    //                 },
    //                 {
    //                     field: "address",
    //                     key: "d",
    //                     title: "Address",
    //                     width: "",
    //                     align: "left",
    //                 },
    //             ],
    //             tableData: TABLE_DATA_SAME_ROW_HEIGHT,
    //             virtualScrollOption: {
    //                 // 是否开启
    //                 enable: true,
    //             },
    //             maxHeight: MAX_HEIGHT,
    //             rowKeyFieldName: "rowKey",
    //         },
    //     });

    //     await later();

    //     wrapper.triggerResizeObserver({ width: MAX_HEIGHT });

    //     await later();

    //     const scrollToFn = mockScrollTo();

    //     wrapper.vm.scrollTo({ top: 1000 });

    //     await later(2000);

    //     // how to test?
    //     expect(wrapper.vm.showVirtualScrollingPlaceholder).toEqual(true);
    // });
});
