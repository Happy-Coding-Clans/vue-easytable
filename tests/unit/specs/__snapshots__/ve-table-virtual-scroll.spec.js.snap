// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`veTable virtual scroll render same row height 1`] = `
<div class="ve-table ve-table-border-around">
  <div class="ve-table-container ve-table-virtual-scroll ve-table-enable-cell-selection" style="max-height: 500px; height: 500px;">
    <div class="ve-table-virtual-phantom ve-table-virtual-scroll" style="height: 40000px;">
      <div style="width: 100%;"></div>
    </div>
    <div class="ve-table-content-wrapper">
      <table class="ve-table-content ve-table-border-x">
        <colgroup>
          <col style="width: 200px;">
          <col style="width: 300px;">
          <col>
        </colgroup>
        <thead class="ve-table-fixed-header ve-table-header">
          <tr class="ve-table-header-tr">
            <th rowspan="1" colspan="1" class="ve-table-header-th" style="text-align: left; top: 0px;">Name</th>
            <th rowspan="1" colspan="1" class="ve-table-header-th" style="text-align: left; top: 0px;">Hobby</th>
            <th rowspan="1" colspan="1" class="ve-table-header-th ve-table-last-column" style="text-align: left; top: 0px;">Address</th>
          </tr>
        </thead>
        <tbody class="ve-table-body ve-table-row-hover ve-table-row-highlight">
          <tr style="height: 0px;">
            <td style="padding: 0px; border: 0px; height: 0px;"></td>
            <td style="padding: 0px; border: 0px; height: 0px;"></td>
            <td style="padding: 0px; border: 0px; height: 0px;"></td>
          </tr>
          <tr row-key="0" class="ve-table-body-tr">
            <td rowspan="1" colspan="1" col-key="b" class="ve-table-body-td" style="text-align: left;">0</td>
            <td rowspan="1" colspan="1" col-key="c" class="ve-table-body-td" style="text-align: left;">0</td>
            <td rowspan="1" colspan="1" col-key="d" class="ve-table-body-td" style="text-align: left;">0</td>
          </tr>
          <tr row-key="1" class="ve-table-body-tr">
            <td rowspan="1" colspan="1" col-key="b" class="ve-table-body-td" style="text-align: left;">1</td>
            <td rowspan="1" colspan="1" col-key="c" class="ve-table-body-td" style="text-align: left;">1</td>
            <td rowspan="1" colspan="1" col-key="d" class="ve-table-body-td" style="text-align: left;">1</td>
          </tr>
          <tr row-key="2" class="ve-table-body-tr">
            <td rowspan="1" colspan="1" col-key="b" class="ve-table-body-td" style="text-align: left;">2</td>
            <td rowspan="1" colspan="1" col-key="c" class="ve-table-body-td" style="text-align: left;">2</td>
            <td rowspan="1" colspan="1" col-key="d" class="ve-table-body-td" style="text-align: left;">2</td>
          </tr>
          <tr row-key="3" class="ve-table-body-tr">
            <td rowspan="1" colspan="1" col-key="b" class="ve-table-body-td" style="text-align: left;">3</td>
            <td rowspan="1" colspan="1" col-key="c" class="ve-table-body-td" style="text-align: left;">3</td>
            <td rowspan="1" colspan="1" col-key="d" class="ve-table-body-td" style="text-align: left;">3</td>
          </tr>
          <tr row-key="4" class="ve-table-body-tr">
            <td rowspan="1" colspan="1" col-key="b" class="ve-table-body-td" style="text-align: left;">4</td>
            <td rowspan="1" colspan="1" col-key="c" class="ve-table-body-td" style="text-align: left;">4</td>
            <td rowspan="1" colspan="1" col-key="d" class="ve-table-body-td" style="text-align: left;">4</td>
          </tr>
          <tr row-key="5" class="ve-table-body-tr">
            <td rowspan="1" colspan="1" col-key="b" class="ve-table-body-td" style="text-align: left;">5</td>
            <td rowspan="1" colspan="1" col-key="c" class="ve-table-body-td" style="text-align: left;">5</td>
            <td rowspan="1" colspan="1" col-key="d" class="ve-table-body-td" style="text-align: left;">5</td>
          </tr>
          <tr row-key="6" class="ve-table-body-tr">
            <td rowspan="1" colspan="1" col-key="b" class="ve-table-body-td" style="text-align: left;">6</td>
            <td rowspan="1" colspan="1" col-key="c" class="ve-table-body-td" style="text-align: left;">6</td>
            <td rowspan="1" colspan="1" col-key="d" class="ve-table-body-td" style="text-align: left;">6</td>
          </tr>
          <tr row-key="7" class="ve-table-body-tr">
            <td rowspan="1" colspan="1" col-key="b" class="ve-table-body-td" style="text-align: left;">7</td>
            <td rowspan="1" colspan="1" col-key="c" class="ve-table-body-td" style="text-align: left;">7</td>
            <td rowspan="1" colspan="1" col-key="d" class="ve-table-body-td" style="text-align: left;">7</td>
          </tr>
          <tr row-key="8" class="ve-table-body-tr">
            <td rowspan="1" colspan="1" col-key="b" class="ve-table-body-td" style="text-align: left;">8</td>
            <td rowspan="1" colspan="1" col-key="c" class="ve-table-body-td" style="text-align: left;">8</td>
            <td rowspan="1" colspan="1" col-key="d" class="ve-table-body-td" style="text-align: left;">8</td>
          </tr>
          <tr row-key="9" class="ve-table-body-tr">
            <td rowspan="1" colspan="1" col-key="b" class="ve-table-body-td" style="text-align: left;">9</td>
            <td rowspan="1" colspan="1" col-key="c" class="ve-table-body-td" style="text-align: left;">9</td>
            <td rowspan="1" colspan="1" col-key="d" class="ve-table-body-td" style="text-align: left;">9</td>
          </tr>
          <tr row-key="10" class="ve-table-body-tr">
            <td rowspan="1" colspan="1" col-key="b" class="ve-table-body-td" style="text-align: left;">10</td>
            <td rowspan="1" colspan="1" col-key="c" class="ve-table-body-td" style="text-align: left;">10</td>
            <td rowspan="1" colspan="1" col-key="d" class="ve-table-body-td" style="text-align: left;">10</td>
          </tr>
          <tr row-key="11" class="ve-table-body-tr">
            <td rowspan="1" colspan="1" col-key="b" class="ve-table-body-td" style="text-align: left;">11</td>
            <td rowspan="1" colspan="1" col-key="c" class="ve-table-body-td" style="text-align: left;">11</td>
            <td rowspan="1" colspan="1" col-key="d" class="ve-table-body-td" style="text-align: left;">11</td>
          </tr>
          <tr row-key="12" class="ve-table-body-tr">
            <td rowspan="1" colspan="1" col-key="b" class="ve-table-body-td" style="text-align: left;">12</td>
            <td rowspan="1" colspan="1" col-key="c" class="ve-table-body-td" style="text-align: left;">12</td>
            <td rowspan="1" colspan="1" col-key="d" class="ve-table-body-td" style="text-align: left;">12</td>
          </tr>
          <tr row-key="13" class="ve-table-body-tr">
            <td rowspan="1" colspan="1" col-key="b" class="ve-table-body-td" style="text-align: left;">13</td>
            <td rowspan="1" colspan="1" col-key="c" class="ve-table-body-td" style="text-align: left;">13</td>
            <td rowspan="1" colspan="1" col-key="d" class="ve-table-body-td" style="text-align: left;">13</td>
          </tr>
          <tr row-key="14" class="ve-table-body-tr">
            <td rowspan="1" colspan="1" col-key="b" class="ve-table-body-td" style="text-align: left;">14</td>
            <td rowspan="1" colspan="1" col-key="c" class="ve-table-body-td" style="text-align: left;">14</td>
            <td rowspan="1" colspan="1" col-key="d" class="ve-table-body-td" style="text-align: left;">14</td>
          </tr>
          <tr row-key="15" class="ve-table-body-tr">
            <td rowspan="1" colspan="1" col-key="b" class="ve-table-body-td" style="text-align: left;">15</td>
            <td rowspan="1" colspan="1" col-key="c" class="ve-table-body-td" style="text-align: left;">15</td>
            <td rowspan="1" colspan="1" col-key="d" class="ve-table-body-td" style="text-align: left;">15</td>
          </tr>
          <tr row-key="16" class="ve-table-body-tr">
            <td rowspan="1" colspan="1" col-key="b" class="ve-table-body-td" style="text-align: left;">16</td>
            <td rowspan="1" colspan="1" col-key="c" class="ve-table-body-td" style="text-align: left;">16</td>
            <td rowspan="1" colspan="1" col-key="d" class="ve-table-body-td" style="text-align: left;">16</td>
          </tr>
          <tr row-key="17" class="ve-table-body-tr">
            <td rowspan="1" colspan="1" col-key="b" class="ve-table-body-td" style="text-align: left;">17</td>
            <td rowspan="1" colspan="1" col-key="c" class="ve-table-body-td" style="text-align: left;">17</td>
            <td rowspan="1" colspan="1" col-key="d" class="ve-table-body-td" style="text-align: left;">17</td>
          </tr>
          <tr row-key="18" class="ve-table-body-tr">
            <td rowspan="1" colspan="1" col-key="b" class="ve-table-body-td" style="text-align: left;">18</td>
            <td rowspan="1" colspan="1" col-key="c" class="ve-table-body-td" style="text-align: left;">18</td>
            <td rowspan="1" colspan="1" col-key="d" class="ve-table-body-td" style="text-align: left;">18</td>
          </tr>
          <tr row-key="19" class="ve-table-body-tr">
            <td rowspan="1" colspan="1" col-key="b" class="ve-table-body-td" style="text-align: left;">19</td>
            <td rowspan="1" colspan="1" col-key="c" class="ve-table-body-td" style="text-align: left;">19</td>
            <td rowspan="1" colspan="1" col-key="d" class="ve-table-body-td" style="text-align: left;">19</td>
          </tr>
          <tr row-key="20" class="ve-table-body-tr">
            <td rowspan="1" colspan="1" col-key="b" class="ve-table-body-td" style="text-align: left;">20</td>
            <td rowspan="1" colspan="1" col-key="c" class="ve-table-body-td" style="text-align: left;">20</td>
            <td rowspan="1" colspan="1" col-key="d" class="ve-table-body-td" style="text-align: left;">20</td>
          </tr>
          <tr row-key="21" class="ve-table-body-tr">
            <td rowspan="1" colspan="1" col-key="b" class="ve-table-body-td" style="text-align: left;">21</td>
            <td rowspan="1" colspan="1" col-key="c" class="ve-table-body-td" style="text-align: left;">21</td>
            <td rowspan="1" colspan="1" col-key="d" class="ve-table-body-td" style="text-align: left;">21</td>
          </tr>
          <tr row-key="22" class="ve-table-body-tr">
            <td rowspan="1" colspan="1" col-key="b" class="ve-table-body-td" style="text-align: left;">22</td>
            <td rowspan="1" colspan="1" col-key="c" class="ve-table-body-td" style="text-align: left;">22</td>
            <td rowspan="1" colspan="1" col-key="d" class="ve-table-body-td" style="text-align: left;">22</td>
          </tr>
          <tr row-key="23" class="ve-table-body-tr">
            <td rowspan="1" colspan="1" col-key="b" class="ve-table-body-td" style="text-align: left;">23</td>
            <td rowspan="1" colspan="1" col-key="c" class="ve-table-body-td" style="text-align: left;">23</td>
            <td rowspan="1" colspan="1" col-key="d" class="ve-table-body-td" style="text-align: left;">23</td>
          </tr>
          <tr row-key="24" class="ve-table-body-tr">
            <td rowspan="1" colspan="1" col-key="b" class="ve-table-body-td" style="text-align: left;">24</td>
            <td rowspan="1" colspan="1" col-key="c" class="ve-table-body-td" style="text-align: left;">24</td>
            <td rowspan="1" colspan="1" col-key="d" class="ve-table-body-td" style="text-align: left;">24</td>
          </tr>
          <tr row-key="25" class="ve-table-body-tr">
            <td rowspan="1" colspan="1" col-key="b" class="ve-table-body-td" style="text-align: left;">25</td>
            <td rowspan="1" colspan="1" col-key="c" class="ve-table-body-td" style="text-align: left;">25</td>
            <td rowspan="1" colspan="1" col-key="d" class="ve-table-body-td" style="text-align: left;">25</td>
          </tr>
        </tbody>
        <tfoot class="ve-table-fixed-footer ve-table-footer"></tfoot>
      </table>
      <div class="ve-table-selection-wrapper">
        <div class="ve-table-selection-fixed-left"></div>
        <div class="ve-table-selection-middle"></div>
        <div class="ve-table-selection-fixed-right"></div>
      </div>
    </div>
  </div>
  <div class="ve-table-edit-input-container" style="top: 0px; left: 0px; height: 1px; z-index: -1; opacity: 0;"><textarea tabindex="-1" class="ve-table-edit-input"></textarea></div>
</div>
`;

exports[`veTable virtual scroll virtual scroll palceholder 1`] = `
<tr class="ve-table-body-tr ve-table-body-row-scrolling">
  <td colspan="1"></td>
</tr>
`;
